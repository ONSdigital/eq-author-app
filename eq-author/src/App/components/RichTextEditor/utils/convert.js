import { convertToHTML, convertFromHTML } from "draft-convert";
import { EditorState } from "draft-js";

export const toHTML = entityMap => {
  const entityToHTML = (entity, originalText) => {
    const mapper = entityMap[entity.type];
    return mapper ? mapper(entity) : originalText;
  };

  const convert = convertToHTML({ entityToHTML });

  return editorState => convert(editorState.getCurrentContent());
};

export const fromHTML = htmlToEntity => {
  const convert = convertFromHTML({ htmlToEntity });

  return (html, decorator) =>
    EditorState.createWithContent(convert(html), decorator);
};
