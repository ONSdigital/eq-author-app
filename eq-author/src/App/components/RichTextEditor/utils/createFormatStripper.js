import { mapKeys } from "lodash";
import { filterEditorState } from "draftjs-filters";
import { filterConfig } from "App/components/RichTextEditor/entities/PipedValue";

const mapper = {
  bold: { format: "BOLD", type: "styles" },
  emphasis: { format: "ITALIC", type: "styles" },
  list: { format: "unordered-list-item", type: "blocks" },
  heading: { format: "header-two", type: "blocks" }
};

export default function createFormatStripper(controls) {
  const filterConfiguration = {
    blocks: [],
    styles: [],
    entities: [filterConfig],
    maxNesting: 0,
    whitespacedCharacters: []
  };

  mapKeys(controls, (value, key) => {
    if (mapper[key] && controls[key]) {
      filterConfiguration[mapper[key].type].push(mapper[key].format);
    }
  });

  return editorState => filterEditorState(filterConfiguration, editorState);
}
