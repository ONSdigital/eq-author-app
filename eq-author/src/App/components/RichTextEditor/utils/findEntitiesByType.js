export default entityType => {
  return function findEntitiesByType(contentBlock, callback, contentState) {
    contentBlock.findEntityRanges(character => {
      const entityKey = character.getEntity();
      return (
        character.getEntity() !== null &&
        contentState.getEntity(entityKey).getType() === entityType
      );
    }, callback);
  };
};
