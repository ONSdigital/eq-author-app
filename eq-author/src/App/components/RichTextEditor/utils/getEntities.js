function getEntities(contentState, entityType) {
  const entities = [];

  contentState.getBlocksAsArray().forEach(block => {
    let selectedEntity = null;

    block.findEntityRanges(
      character => {
        const entityKey = character.getEntity();

        if (entityKey === null) {
          return false;
        }

        const entity = contentState.getEntity(entityKey);

        if (entityType && entity.getType() !== entityType) {
          return false;
        }

        selectedEntity = {
          entityKey,
          entity,
          blockKey: block.getKey()
        };

        return true;
      },
      (start, end) => {
        entities.push({ ...selectedEntity, start, end });
      }
    );
  });

  return entities;
}

export default getEntities;
