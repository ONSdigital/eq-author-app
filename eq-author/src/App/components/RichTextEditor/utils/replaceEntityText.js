import { SelectionState, Modifier } from "draft-js";

const getEntityRange = (block, entityKey) => {
  let range;

  // get the range for the entity key
  block.findEntityRanges(
    char => char.getEntity() === entityKey,
    (start, end) => {
      range = { start, end };
    }
  );

  return SelectionState.createEmpty(block.key).merge({
    anchorKey: block.key,
    anchorOffset: range.start,
    focusKey: block.key,
    focusOffset: range.end
  });
};

const replaceText = (contentState, entityKey, blockKey, text, inlineStyles) => {
  const block = contentState.getBlockForKey(blockKey);
  const entitySelection = getEntityRange(block, entityKey);

  // first replace text
  const updatedContentState = Modifier.replaceText(
    contentState,
    entitySelection,
    text,
    inlineStyles,
    entityKey
  );

  // then create new selection, which covers the updated entity
  const updatedEntitySelection = entitySelection.merge({
    focusOffset: entitySelection.getStartOffset() + text.length
  });

  // apply the updated entity to the content
  return Modifier.applyEntity(
    updatedContentState,
    updatedEntitySelection,
    entityKey
  );
};

export default replaceText;
