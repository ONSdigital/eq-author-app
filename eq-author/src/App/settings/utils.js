export const getThemeSettingsErrorCount = (questionnaire) =>
  questionnaire?.themeSettings?.validationErrorInfo?.totalCount +
    questionnaire?.validationErrorInfo?.totalCount || 0;
