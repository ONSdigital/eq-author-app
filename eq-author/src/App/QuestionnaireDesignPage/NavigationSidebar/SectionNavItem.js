import gql from "graphql-tag";
import React from "react";
import styled from "styled-components";
import { withRouter } from "react-router-dom";

import CustomPropTypes from "custom-prop-types";
import { colors } from "constants/theme";
import PageNav from "./PageNav";
import NavLink from "./NavLink";
import { buildSectionPath } from "utils/UrlUtils";
import SectionIcon from "./icon-section.svg?inline";

const StyledSectionNavItem = styled.li`
  display: block;
`;

const SectionNameOuter = styled.span`
  padding-left: 0.5em;
`;

const StyledSectionUpper = styled.div`
  display: block;
  padding-top: 1px;
  margin-left: 0.75em;
  border-top: 1px solid ${colors.grey};
`;

const StyledSectionLower = styled.div`
  display: block;
  padding-bottom: 1px;
  margin-bottom: 1px;
  margin-left: 0.75em;
  border-bottom: 1px solid ${colors.grey};
`;

export class UnwrappedSectionNavItem extends React.Component {
  static propTypes = {
    questionnaire: CustomPropTypes.questionnaire,
    section: CustomPropTypes.section.isRequired,
    match: CustomPropTypes.match.isRequired,
  };

  render() {
    const { questionnaire, section, match, ...otherProps } = this.props;

    const url = buildSectionPath({
      questionnaireId: questionnaire.id,
      sectionId: section.id,
      tab: match.params.tab,
    });

    return (
      <StyledSectionNavItem data-test="section-item" {...otherProps}>
        <StyledSectionUpper>
          <div />
        </StyledSectionUpper>
        <NavLink
          exact
          to={url}
          data-test="nav-section-link"
          title={section.displayName}
          icon={SectionIcon}
          id="sectionName"
          errorCount={section.validationErrorInfo.totalCount}
        >
          <SectionNameOuter>{section.displayName}</SectionNameOuter>
        </NavLink>
        <StyledSectionLower>
          <div />
        </StyledSectionLower>

        <PageNav section={section} questionnaire={questionnaire} />
      </StyledSectionNavItem>
    );
  }
}

UnwrappedSectionNavItem.fragments = {
  SectionNavItem: gql`
    fragment SectionNavItem on Section {
      id
      title
      displayName
      questionnaire {
        id
      }
      validationErrorInfo {
        id
        totalCount
      }
      ...PageNav
    }

    ${PageNav.fragments.PageNav}
  `,
};

export default withRouter(UnwrappedSectionNavItem);
