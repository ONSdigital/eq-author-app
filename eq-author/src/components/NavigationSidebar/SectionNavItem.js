import React from "react";
import styled from "styled-components";

import PageNav from "components/NavigationSidebar/PageNav";
import NavLink from "./NavLink";
import CustomPropTypes from "custom-prop-types";
import { buildSectionPath } from "utils/UrlUtils";
import SectionIcon from "./icon-section.svg?inline";
import gql from "graphql-tag";

const StyledSectionNavItem = styled.li`
  display: block;
`;

class SectionNavItem extends React.Component {
  static propTypes = {
    questionnaire: CustomPropTypes.questionnaire,
    section: CustomPropTypes.section.isRequired
  };

  render() {
    const { questionnaire, section, ...otherProps } = this.props;
    const url = buildSectionPath({
      questionnaireId: questionnaire.id,
      sectionId: section.id,
      tab: "design"
    });

    return (
      <StyledSectionNavItem data-test="section-item" {...otherProps}>
        <NavLink
          exact
          to={url}
          data-test="nav-section-link"
          title={section.displayName}
          icon={SectionIcon}
        >
          {section.displayName}
        </NavLink>

        <PageNav section={section} questionnaire={questionnaire} />
      </StyledSectionNavItem>
    );
  }
}

SectionNavItem.fragments = {
  SectionNavItem: gql`
    fragment SectionNavItem on Section {
      id
      title
      displayName
      ...PageNav
    }

    ${PageNav.fragments.PageNav}
  `
};

export default SectionNavItem;
