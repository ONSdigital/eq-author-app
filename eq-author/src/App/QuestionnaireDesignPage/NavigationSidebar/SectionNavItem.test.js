import { shallow } from "enzyme";
import React from "react";
import { buildQuestionnaire } from "tests/utils/createMockQuestionnaire";

import { UnwrappedSectionNavItem as SectionNavItem } from "./SectionNavItem";

describe("SectionNavItem", () => {
  const questionnaire = buildQuestionnaire();

  it("should render", () => {
    const handleAddPage = jest.fn(() => Promise.resolve);
    const wrapper = shallow(
      <SectionNavItem
        questionnaire={questionnaire}
        section={questionnaire.sections[0]}
        onAddPage={handleAddPage}
        duration={123}
        isActive={jest.fn()}
        isOpen={{ open: true }}
        handleChange={jest.fn()}
        identity={1}
        match={{
          params: {
            questionnaireId: questionnaire.id,
            tab: "design",
          },
        }}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
});
