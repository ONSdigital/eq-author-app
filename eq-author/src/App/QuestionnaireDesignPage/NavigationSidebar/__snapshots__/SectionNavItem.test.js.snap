// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SectionNavItem should render 1`] = `
<SectionNavItem__StyledSectionNavItem
  data-test="section-item"
  duration={123}
  isActive={[MockFunction]}
  onAddPage={[MockFunction]}
>
  <SectionNavItem__StyledSectionUpper>
    <div />
  </SectionNavItem__StyledSectionUpper>
  <NavLink
    data-test="nav-section-link"
    errorCount={2}
    exact={true}
    icon={[Function]}
    id="sectionName"
    title="Section"
    to="/q/1/section/3/design"
  >
    <SectionNavItem__SectionNameOuter>
      Section
    </SectionNavItem__SectionNameOuter>
  </NavLink>
  <SectionNavItem__StyledSectionLower>
    <div />
  </SectionNavItem__StyledSectionLower>
  <PageNav
    questionnaire={
      Object {
        "displayName": "Questionnaire",
        "id": "1",
        "sections": Array [
          Object {
            "displayName": "Section",
            "id": "3",
            "pages": Array [
              Object {
                "displayName": "Page",
                "id": "2",
                "title": "Page",
              },
            ],
            "title": "Section",
            "validationErrorInfo": Object {
              "totalCount": 2,
            },
          },
        ],
        "title": "Questionnaire",
      }
    }
    section={
      Object {
        "displayName": "Section",
        "id": "3",
        "pages": Array [
          Object {
            "displayName": "Page",
            "id": "2",
            "title": "Page",
          },
        ],
        "title": "Section",
        "validationErrorInfo": Object {
          "totalCount": 2,
        },
      }
    }
  />
</SectionNavItem__StyledSectionNavItem>
`;
